src/main/haskell/language-kore/src/Data/Kore/Predicate/Predicate.hs

```diff
@@ -0,0 +1,322 @@
+-}
+module Data.Kore.Predicate.Predicate
+    ( PredicateProof (..)
+    , CommonPredicate -- Constructor not exported on purpose
+    , Predicate -- Constructor not exported on purpose
+    , pattern PredicateFalse
+    , pattern PredicateTrue
+    , compactPredicatePredicate
+    , makeAndPredicate
+    , makeMultipleAndPredicate
+    , makeCeilPredicate
+    , makeEqualsPredicate
+    , makeFalsePredicate
+    , makeIffPredicate
+    , makeImpliesPredicate
+    , makeNotPredicate
+    , makeOrPredicate
+    , makeTruePredicate
+    , stringFromPredicate
+    , unwrapPredicate
+    , variableSetFromPredicate
+    , wrapPredicate
+    ) where
+
+import           Data.Fix                              (cata)
+import           Data.List                             (foldl')
+import           Data.Reflection                       (Given)
+import qualified Data.Set                              as Set
+
+import           Data.Kore.AST.Common                  (And (..), Exists (..),
+                                                        Forall (..), Iff (..),
+                                                        Implies (..), Not (..),
+                                                        Or (..), Pattern (..),
+                                                        SortedVariable,
+                                                        Variable)
+import           Data.Kore.AST.MetaOrObject
+import           Data.Kore.AST.PureML                  (PureMLPattern)
+import           Data.Kore.ASTUtils.SmartConstructors  (pattern Bottom_,
+                                                        pattern Top_, mkAnd,
+                                                        mkBottom, mkCeil,
+                                                        mkEquals, mkIff,
+                                                        mkImplies, mkNot, mkOr,
+                                                        mkTop)
+import           Data.Kore.IndexedModule.MetadataTools (MetadataTools)
+
+{--| 'PredicateProof' is a placeholder for a proof showing that a Predicate
+evaluation was correct.
+--}
+data PredicateProof level = PredicateProof
+    deriving (Show, Eq)
+
+{--| 'GenericPredicate' is a wrapper for predicates used for type safety.
+Should not be exported, and should be treated as an opaque entity which
+can be manipulateds only by functions in this module.
+--}
+newtype GenericPredicate pat = GenericPredicate pat
+    deriving (Show, Eq, Functor, Traversable, Foldable)
+
+{--| 'Predicate' is a user-visible representation for predicates.
+--}
+type Predicate level var = GenericPredicate (PureMLPattern level var)
+
+{--| 'CommonPredicate' follows the generic convention of particularizing types
+to Variable.
+--}
+type CommonPredicate level = Predicate level Variable
+
+{- 'compactPredicatePredicate' removes one level of 'GenericPredicate' which
+sometimes occurs when, say, using Predicates as Traversable.
+-}
+compactPredicatePredicate
```

Comment:
This is an instance of join from Monad;  and the way you use it seems to be similar to how join is used...

Reply:
I wouldn't make this a Monad just to have a nice join.